app/src/main/java/com/android/sample/model/user/UserRepositoryFirestore.kt

```diff
@@ -0,0 +1,86 @@
+package com.android.sample.model.user
+
+import android.util.Log
+import com.google.android.gms.tasks.Task
+import com.google.firebase.Firebase
+import com.google.firebase.auth.auth
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+
+class UserRepositoryFirestore(private val db: FirebaseFirestore) : UserRepository {
+  private val collectionPath = "users"
+
+  override fun init(onSuccess: () -> Unit) {
+    if (Firebase.auth.currentUser != null) {
+      onSuccess()
+    }
+  }
+
+  override fun getUserById(id: String, onSuccess: (User) -> Unit, onFailure: (Exception) -> Unit) {
+    val doc = db.collection(collectionPath).document(id)
+    doc.get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        onSuccess(convertSnapshot(id, task.result))
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun addUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performUserDatabaseAction(
+        db.collection(collectionPath).document(user.uid).set(user), onSuccess, onFailure)
+  }
+
+  override fun updateUser(user: User, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performUserDatabaseAction(
+        db.collection(collectionPath).document(user.uid).set(user), onSuccess, onFailure)
+  }
+
+  override fun deleteUserById(id: String, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    performUserDatabaseAction(
+        db.collection(collectionPath).document(id).delete(), onSuccess, onFailure)
+  }
+
+  /**
+   * Method that allows us to convert DocumentSnapshots into User data classes
+   *
+   * @param uid unique id of the user of which the snapshot represents
+   * @param snapshot DocumentSnapshot that was extracted from Firestore and which we want to convert
+   *   into a User
+   * @return a [User] data class that is created from the DocumentSnapshot
+   */
+  @Suppress("UNCHECKED_CAST")
+  private fun convertSnapshot(uid: String, snapshot: DocumentSnapshot): User {
+    val userName = snapshot["userName"] as String
```

Comment:
The `@Suppress("UNCHECKED_CAST")` is there to remove the warnings from the cast `val likedRecipes = snapshot["likedRecipes"] as List<String>`. I decided to remove it because when we are storing the `fridge`, `likedRecipes` and `createdRecipes` as a `List<String>`. Thus, when we extract it from the database, we are guaranteed that we get a `List<String>`.
Concerning the `val userName = snapshot["userName"] as String`, it is like this because there will be always data stored in this field. Thus, we will never get `null`.